hail/src/main/scala/is/hail/expr/ir/InferPType.scala

```diff
@@ -226,27 +303,31 @@ object InferPType {
       }
       case NDArrayRef(nd, idxs) => {
```

Comment:
I think this is wrong. I think we should check whether each field referenced by idx's are required.

Reply:
I think I'm a little confused by what you're suggesting. Aren't we already enforcing that the type of this node is required if all of the idx types are required? Are you suggesting something else?

Reply:
I think NDArrayRef is used (downstream, in codegen) to fetch certain elements of an NDArray, e.g a view of the backing array. Even if the backing NDArray has missing elements, the view may not.

Reply:
The backing NDArray isn't allowed to have missing elements, I think (currently a runtime check because we don't handle requiredness well). The element referenced will be missing if any of the indices are missing; otherwise it will be present.

Reply:
But the actual `NDArray` value only exists at run-time (and the indices as well), how can we tell now, at compile-time, whether the view happens to contain any missing elements? The only thing we can say now is: if the elements of the NDArray might be missing, then it is possible for some elements in this view to be missing. If the elements of the NDArray are guaranteed (statically) to not be missing, then we can make the same guarantee of any view into the NDArray.

Reply:
Depends on whether the NDArray values are known at compile time. Also, we could do this at a granularity smaller than the entire NDArray: for instance I could image a feature that I know will always be present (a particular column) for every sample, and want this knowledge to be encoded into the type.

Reply:
NDArrays are homogeneous structures. The elements are all the same type. Tracking requiredness for particular columns is more of a table feature than an NDArray feature. We don't even know statically how many columns an NDArray has!

But also, Arcturus is right. We don't allow NDArrays to have missing elements anyways, to enable efficient dense linear algebra implementations.

> Depends on whether the NDArray values are known at compile time.

I think for the most part, making requiredness decisions based on statically known values should be handled by the constant folder, not the ptype inference.